ibm/resource_ibm_ipsec_vpn.go

```diff
@@ -0,0 +1,537 @@
+	IPSecOrder := datatypes.Container_Product_Order_Network_Tunnel_Ipsec{
+		Container_Product_Order: datatypes.Container_Product_Order{
+			PackageId: sl.Int(packageid),
+			Prices:    priceItems,
+			Quantity:  sl.Int(1),
+			Location:  &locationid,
+		},
+	}
+	//Calling verify order
+	_, err = services.GetProductOrderService(sess.SetRetries(0)).
+		VerifyOrder(&IPSecOrder)
+	if err != nil {
+		return fmt.Errorf("Error during Verify order for Creating: %s", err)
+	}
+
+	//Calling place order
+	receipt, err := services.GetProductOrderService(sess.SetRetries(0)).
+		PlaceOrder(&IPSecOrder, sl.Bool(false))
+	if err != nil {
+		return fmt.Errorf("Error during Place order for Creating: %s", err)
+	}
+	vpn, _ := findIPSecVpnByOrderID(sess, *receipt.OrderId, d)
+	if err != nil {
+		return fmt.Errorf("Error during creation of IPSec VPN: %s", err)
+	}
+	id := *vpn.Id
+	d.SetId(fmt.Sprintf("%d", id))
+	log.Printf("[INFO] IPSec VPN ID: %s", d.Id())
+	return resourceIBMIPSecVPNUpdate(d, meta)
+}
+
+func findIPSecVpnByOrderID(sess *session.Session, orderID int, d *schema.ResourceData) (datatypes.Network_Tunnel_Module_Context, error) {
+	filterPath := "networkTunnelContexts.billingItem.orderItem.order.id"
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{"pending"},
+		Target:  []string{"complete"},
+		Refresh: func() (interface{}, string, error) {
+			vpn, err := services.GetAccountService(sess).
+				Filter(filter.Build(
+					filter.Path(filterPath).
+						Eq(strconv.Itoa(orderID)))).
+				Mask(ipsecMask).
+				GetNetworkTunnelContexts()
+			if err != nil {
+				return datatypes.Network_Tunnel_Module_Context{}, "", err
+			}
+
+			if len(vpn) == 1 {
+				return vpn[0], "complete", nil
+			} else if len(vpn) == 0 {
+				return nil, "pending", nil
+			}
+			return nil, "", fmt.Errorf("Expected one IPSec VPN: %s", err)
+		},
+		Timeout:        2 * time.Hour,
+		Delay:          10 * time.Second,
+		MinTimeout:     10 * time.Second,
+		NotFoundChecks: 24 * 60,
+	}
+
+	pendingResult, err := stateConf.WaitForState()
+
+	if err != nil {
+		return datatypes.Network_Tunnel_Module_Context{}, err
+	}
+	var result, ok = pendingResult.(datatypes.Network_Tunnel_Module_Context)
+	if ok {
+		return result, nil
+	}
+
+	return datatypes.Network_Tunnel_Module_Context{},
+		fmt.Errorf("Cannot find IPSec Vpn with order id '%d'", orderID)
+}
+
+func resourceIBMIPSecVPNRead(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(ClientSession).SoftLayerSession()
+	vpnID, _ := strconv.Atoi(d.Id())
+
+	vpn, err := services.GetNetworkTunnelModuleContextService(sess).
+		Id(vpnID).Mask(ipsecMask).
+		GetObject()
+	if err != nil {
+		return fmt.Errorf("Error retrieving firewall information: %s", err)
+	}
+	d.Set("name", *vpn.Name)
+	d.Set("internal_peer_ip_address", *vpn.InternalPeerIpAddress)
+	if vpn.Datacenter != nil {
+		d.Set("datacenter", *vpn.Datacenter.Name)
+	}
+	phaseoneAttributesMap := make([]map[string]interface{}, 0, 1)
```

Comment:
Changes look fine...small chnage Can you move these attributesMap to respective flatten methods in structures.go (flattenIPSecPhaseOne,flattenIPSecPhaseTwo, flattenIPSecAddressTranslation, flattenIPSecCustomerSubnets)

Reply:
Done. Please review